homeassistant/components/sma/__init__.py

```diff
@@ -1 +1,193 @@
+
+        # Only sensors from config should be enabled
+        for sensor in sensor_def:
+            sensor.enabled = sensor.name in config_sensors
+
+
+async def _migrate_old_unique_ids(
+    hass: HomeAssistant, entry: ConfigEntry, sensor_def: pysma.Sensors
+) -> None:
+    """Migrate legacy sensor entity_id format to new format."""
+    entity_registry = er.async_get(hass)
+
+    # Create list of all possible sensor names
+    possible_sensors = list(
+        set(
+            entry.data.get(CONF_SENSORS)
+            + [s.name for s in sensor_def]
+            + list(pysma.LEGACY_MAP)
+        )
+    )
+
+    for sensor in possible_sensors:
+        if sensor in sensor_def:
+            pysma_sensor = sensor_def[sensor]
+            original_key = pysma_sensor.key
+        elif sensor in pysma.LEGACY_MAP:
+            # If sensor was removed from pysma we will remap it to the new sensor
+            legacy_sensor = pysma.LEGACY_MAP[sensor]
+            pysma_sensor = sensor_def[legacy_sensor["new_sensor"]]
+            original_key = legacy_sensor["old_key"]
+        else:
+            _LOGGER.error("%s does not exist", sensor)
+            continue
+
+        # Find entity_id using previous format of unique ID
+        entity_id = entity_registry.async_get_entity_id(
+            "sensor", "sma", f"sma-{original_key}-{sensor}"
+        )
+
+        if entity_id:
+            # Change entity_id to new format using the device serial in entry.unique_id
+            new_unique_id = (
+                f"{entry.unique_id}-{pysma_sensor.key}_{pysma_sensor.key_idx}"
+            )
+            entity_registry.async_update_entity(entity_id, new_unique_id=new_unique_id)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up sma from a config entry."""
+    # Init all default sensors
+    sensor_def = pysma.Sensors()
+
+    if entry.source == SOURCE_IMPORT:
+        await _parse_legacy_options(entry, sensor_def)
+        await _migrate_old_unique_ids(hass, entry, sensor_def)
+
+    # Init the SMA interface
+    protocol = "https" if entry.data.get(CONF_SSL) else "http"
+    url = f"{protocol}://{entry.data.get(CONF_HOST)}"
+    verify_ssl = entry.data.get(CONF_VERIFY_SSL)
+    group = entry.data.get(CONF_GROUP)
+    password = entry.data.get(CONF_PASSWORD)
+
+    session = async_get_clientsession(hass, verify_ssl=verify_ssl)
+    sma = pysma.SMA(session, url, password, group)
+
+    # Ensure we logout on shutdown
+    async def async_close_session(event):
+        """Close the session."""
+        await sma.close_session()
+
+    remove_stop_listener = hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_STOP, async_close_session
+    )
+
+    # Define the coordinator
+    async def async_update_data():
+        """Update the used SMA sensors."""
+        values = await sma.read(sensor_def)
+        if not values:
+            raise UpdateFailed
```

Comment:
In the future if authentication fails you can `raise ConfigEntryAuthFailed` here and it will start a new flow with `async_step_reauth` in `config_flow.py`

Reply:
Thanks will keep that in mind for the future